packages/language-server/src/ui5-model.ts

```diff
@@ -11,29 +13,81 @@ import { Fetcher } from "../api";
 
 const DEFAULT_UI5_VERSION = "1.71.14";
 
-export async function getSemanticModel(): Promise<UI5SemanticModel> {
-  return getSemanticModelWithFetcher(fetch);
+export async function getSemanticModel(
+  modelCachePath: string | undefined
+): Promise<UI5SemanticModel> {
+  return getSemanticModelWithFetcher(fetch, modelCachePath);
 }
 
 // This function is exported for testing purposes (using a mock fetcher)
 export async function getSemanticModelWithFetcher(
-  fetcher: Fetcher
+  fetcher: Fetcher,
+  modelCachePath: string | undefined
 ): Promise<UI5SemanticModel> {
   const version = DEFAULT_UI5_VERSION;
   const jsonMap: Record<string, Json> = {};
   const baseUrl = `https://sapui5.hana.ondemand.com/${version}/test-resources/`;
   const suffix = "/designtime/api.json";
   const libs = getLibs();
+  let cacheFolder: string | undefined;
+
+  // Note: all cache handling (reading, writing etc) is optional from the user perspective but
+  // impacts performance, therefore if any errors occur when handling the cache we ignore them but output
+  // a warning to the user
+  if (modelCachePath !== undefined) {
+    cacheFolder = getCacheFolder(modelCachePath, version);
+    try {
+      await mkdirs(cacheFolder);
+    } catch (err) {
+      console.warn(
+        `Could not create folder ${cacheFolder} for caching UI5 resources`,
+        err
+      );
+      cacheFolder = undefined;
+    }
+  }
 
   await Promise.all(
     map(libs, async libName => {
-      const url = baseUrl + libName + suffix;
-      const response = await fetcher(url);
-      if (response.ok) {
-        const apiJson = await response.json();
+      let filePath;
+      let apiJson;
+      if (cacheFolder !== undefined) {
+        filePath = getCacheFilePath(cacheFolder, libName);
```

Comment:
could this be flow be extracted to two sub functions?
- Cache One
- Network one.

Reply:
Not really because if we couldn't read from the cache we'll do the network call anyway, but I'll extract the cache handling to 2 functions so the flow is clearer